kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go

```diff
@@ -0,0 +1,188 @@
+	if *classCollectorMetrics == allMetrics {
+		for _, m := range allClassMetrics {
+			c.metrics[m] = struct{}{}
+		}
+	} else if *classCollectorMetrics != noMetrics {
+		for _, m := range strings.Split(*classCollectorMetrics, ",") {
+			c.metrics[m] = struct{}{}
+		}
+	}
+	c.Start()
+	return c, nil
+}
+
+func (c *classCollector) Start() {
+	c.logger.Log("info", "start class collector", "metrics", getMetricsKeys(c.metrics))
+	if isNeedCollectMbLLc(c.metrics) {
+		go func() {
+			for {
+				err := c.updateClasses()
+				if err != nil {
+					c.logger.Log("error", "class collector update classes failed", "err", err)
+				}
+				time.Sleep(jitter(c.interval))
+			}
+		}()
+	}
+	go func() {
+		for {
+			err := c.updateStats()
+			if err != nil {
+				c.logger.Log("error", "class collector update stats failed", "err", err)
+			}
+			time.Sleep(jitter(c.interval))
+		}
+	}()
+}
+
+func (c *classCollector) updateClasses() error {
+	excludeDirs := map[string]bool{
+		"info":       true,
+		"mon_data":   true,
+		"mon_groups": true,
+	}
+	files, err := os.ReadDir(rootResctrlPath)
+	if err != nil {
+		return err
+	}
+	for _, file := range files {
+		if file.IsDir() {
+			dirName := file.Name()
+			_, ok := c.statsCache[dirName]
+			if !excludeDirs[dirName] && !ok {
+				c.statsCache[dirName] = nil
+			}
+		}
+	}
+	return nil
+}
+
+func (c *classCollector) updateStats() error {
+	for class := range c.statsCache {
+		newStats := RawStats{}
+		var err error
+		if isNeedCollectMbLLc(c.metrics) {
+			newStats.SocketNum, newStats.MemoryBandwidth, newStats.Cache, err =
+				getIntelRDTStatsFrom(filepath.Join(rootResctrlPath, class))
+			if err != nil {
+				return err
+			}
+		}
+		if c.statsCache[class] != nil {
+			pStats, err := processStats(c.statsCache[class].oldStats, newStats)
+			if err != nil {
+				return err
+			}
+			c.statsCache[class] = &stats{
+				oldStats:       newStats,
+				processedStats: pStats,
+			}
+		} else {
+			c.statsCache[class] = &stats{
+				oldStats:       newStats,
+				processedStats: ProcessedStats{},
+			}
+		}
+	}
+	return nil
+}
+
+func (c *classCollector) Update(ch chan<- prometheus.Metric) error {
+	if len(c.statsCache) == 0 {
+		c.logger.Log("info", "class collector stats have no cache")
+		return nil
+	}
+	if !isNeedCollectMbLLc(c.metrics) {
+		return nil
+	}
+	for cid, stats := range c.statsCache {
+		ch <- prometheus.MustNewConstMetric(
+			classTotalMemoryBandwidthDesc,
+			prometheus.GaugeValue,
+			stats.processedStats.SumMemoryBandwidth.TotalMBps,
+			cid,
+			c.nodeName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			classLocalMemoryBandwidthDesc,
+			prometheus.GaugeValue,
+			stats.processedStats.SumMemoryBandwidth.LocalMBps,
+			cid,
+			c.nodeName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			classLLCacheDesc,
+			prometheus.GaugeValue,
+			stats.processedStats.SumCache.LLCOccupancy,
+			cid,
+			c.nodeName,
+		)
+		for sid, s := range stats.processedStats.MemoryBandwidth {
```

Comment:
a few comments nice .. what is sid? guessing cid is container-id

Reply:
Sorry for the confusion, sid is socket-id and cid is container-id.